_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,791 @@
+        let created_at = Utc::now().format(date_format).to_string();
+        let created_at =
+            DateTime::parse_from_str(&created_at, "%Y/%m/%d %H:%M:%S%.3f %z").unwrap() -
+            chrono::Duration::seconds((number_of_seconds * i).try_into().unwrap());
+
+        if !data.iter().any(|x| x.time == created_at) {
+            data.push(WebsiteStats {
+                time: created_at.into(),
+                uptime_pct: None,
+            });
+        }
+    }
+    data.sort_by(|a, b| b.time.cmp(&a.time));
+}
+
+data
+```
+
+Now we can refactor our handler function:
+
+```rust
+async fn get_websites(State(state): State<AppState>) -> impl AskamaIntoResponse {
+    let websites = sqlx
+        ::query_as::<_, Website>("SELECT url, alias FROM websites")
+        .fetch_all(&state.db).await
+        .unwrap();
+
+    let mut logs = Vec::new();
+
+    for website in websites {
+        let data = get_website_uptime_stats(SplitBy::Hour, &website.alias, &state.db)
+            .await;
+
+        logs.push(WebsiteInfo {
+            url: website.url,
+            alias: website.alias,
+            data,
+        });
+    }
+
+    WebsiteLogs { logs }
+}
+```
+
+Once done, we’ll want to create a dynamic route for getting more information about a monitored URL. We can use this page to display things like past incidents/alerts. This function will follow most of the previous handler function except we’re fetching one URL and additionally grabbing any records of HTTP requests that didn’t return `200 OK` and labelling them as “Incidents”:
+
+```rust
+async fn get_website_by_id(
+    State(state): State<AppState>,
+    Path(alias): Path<String>,
+) -> impl AskamaIntoResponse {
+    let website = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites WHERE alias = $1")
+        .bind(&alias)
+        .fetch_one(&state.db)
+        .await
+        .unwrap();
+
+    // .. rest of your code
+}
+```
+
+Then we can write the previous code for getting the uptime percentages per hour, as well as grabbing a vector of `Incident`s:
+
+```rust
+// main.rs
+let last_24_hours_data = get_website_uptime_stats(SplitBy::Hour, &website.alias, &state.db).await;
+let monthly_data = get_website_uptime_stats(SplitBy::Day, &website.alias, &state.db).await;
+
+let incidents = sqlx::query_as::<_, Incident>(
+        "SELECT created_at as time, status from logs where website_alias = $1
+        and status != 200",
+    )
+    .bind(&alias)
+    .fetch_all(&state.db)
+    .await
+    .unwrap();
+
+    let log = WebsiteInfo {
+        url: website.url,
+        alias,
+        data,
+    };
+
+SingleWebsiteLogs { log, incidents, monthly_data }
+```
+
+“Incidents” in this case are defined as failed requests to a monitored URL. If the SQL query for getting Incidents fails (for example: because there aren’t any reported yet), Askama will automatically turn it into a vector of length zero instead of erroring out and causing a HTTP error.
```

Comment:
I'm not sure how happy I'm with this automatic behaviour. This requires the users, and therefore my team members, to know `Askama` to know what is happening here. I would rather want to code to be explicit and just create an empty vec if sqlx returns no records were found.

As it is now, the query could fail for other reasons and it will silently just become a vec of no incidents.

Reply:
It has also felt like most of the article has so far explained something and then had the code for it. Here the code was presented before the explanation. Not sure if it is intended? :smile:

Reply:
Think this is a mistake. Will go back and fix